### DISCUSSION TITLE
Using the Controls system to change BCs

### DISCUSSION THREAD

**Discussion by mangerij** [idaholab/moose#20208] [Q&A General] [answered]
Hi, I have two questions:

1. If I use a simulation that has not properly ended with a checkpoint or an actual exit as the initial condition for a simulation with a Controls block, do the Controls objects not see the right start_time and end_time?

2. How would I change my BCs block from

```
  [./Periodic]
    [./xyz]
      auto_direction = 'x y z'
      variable = 'mag1_x mag1_y mag1_z mag2_x mag2_y mag2_z'
    [../]
  [../]

  [./boundary_grounding1]
    type = FunctionDirichletBC
    boundary = 'front'
    variable = potential_H_int
    function = pulse
  [../]

  [./boundary_grounding2]
    type = DirichletBC
    boundary = 'back'
    variable = potential_H_int
    value = 0.0
  [../]
```

to

```
  [./Periodic]
    [./xyz]
      auto_direction = 'x y z'
      variable = 'mag1_x mag1_y mag1_z mag2_x mag2_y mag2_z potential_H_int'
    [../]
  [../]
```
with the TimePeriod controls?

I tried a bit and couldn't get it working so I reverted to two input files and I suspect that the first question here is maybe related to why this didn't seem to work.

thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

- The Controls see the simulation time. It does not necessarily match the restart/checkpoint time. When you restart, you specify the timestep that you want to restart from. 

- This should work, please let us know if it doesnt
```
[BCs]
  [./Periodic]
    [./xyz]
      auto_direction = 'x y z'
      variable = 'mag1_x mag1_y mag1_z mag2_x mag2_y mag2_z'
    [../]
    [./xyz_later]
      auto_direction = 'x y z'
      variable = 'mag1_x mag1_y mag1_z mag2_x mag2_y mag2_z potential_H_int'
    [../]
  [../]

  [./boundary_grounding1]
    type = FunctionDirichletBC
    boundary = 'front'
    variable = potential_H_int
    function = pulse
  [../]

  [./boundary_grounding2]
    type = DirichletBC
    boundary = 'back'
    variable = potential_H_int
    value = 0.0
  [../]
[]

[Controls]
  [./diff]
    type = TimePeriod
    enable_objects = 'BCs::Periodic::xyz BCs::boundary_grounding1 BCs::boundary_grounding2'
    disable_objects = 'BCs::Periodic::xyz_later'
    start_time = '<simulation_start_time_with_first_set_of_BCs>'
    end_time = '<time to switch to the other set>'
  [../]
[]
```

**Reply by mangerij** [reply]
Excellent!

**Reply by mangerij** [reply]
When I use that code, I am getting this error from the parser:

*** ERROR ***
The following error occurred in the object "diff", of type "TimePeriod".

The desired parameter 'BCs::Periodic::xyz/enable' was not located for the 'diff' object, it either does not exist or has not been declared as controllable

**Reply by GiudGiud** [reply]
I ll have a look

**Reply by GiudGiud** [reply]
Ok so the periodic boundary conditions are not currently supported in the Control system.
They inherit very fast from libmesh periodic boundary conditions, they dont have most of the moose wrapping the regular BCs have. The [Periodic] action creates them, they are not regular moose objects. 

@andrsd how much work would think adding them to the control system would be? Do we have anyone working on periodic BCs rn who could take a look?

**Reply by mangerij** [reply]
It is not such an important thing for me. More of a matter of convenience. Cycling through two input files is perfectly fine.

I also don't know if immediately making this variable periodic after pulsing the simulation is even the right thing to do :)

thanks!

**Reply by andrsd** [reply]
Periodic BCs change the underlying system in libMesh, so if you want to change those via TimePeriod, then you have some non-trivial work to do with re-projecting solutions, renumbering DoF, etc.  That would be my guess...
